File: py/2022/day15.py
from dataclasses import dataclass
from itertools import product
from typing import Final

from aoc import main
from aoc.coords import Point, mdist
from aoc.parse import all_numbers

FREQUENCY_MULTIPLIER: Final[int] = 4000000


def parse(s: str) -> tuple[dict[Point, int], list[Point]]:
    sensors = {}
    beacons = []
    for line in s.splitlines():
        (sx, sy, bx, by) = all_numbers(line)
        sensor = (sx, sy)
        beacon = (bx, by)
        sensors[sensor] = mdist(sensor, beacon)
        beacons.append(beacon)
    return sensors, beacons


@dataclass(eq=True)
class Range:
    start: int
    end: int

    @staticmethod
    def union(*ranges: "Range") -> "list[Range]":
        union = []
        last = None
        for x in sorted(ranges):
            if last and last.end >= x.start - 1:
                last.end = max(last.end, x.end)
            else:
                last = x
                union.append(last)
        return union

    def __lt__(self, other: "Range") -> bool:
        return self.start < other.start

    def __len__(self) -> int:
        return self.end - self.start + 1

    def __contains__(self, other: int) -> bool:
        return self.start <= other <= self.end

    def overlaps(self, other: "Range") -> bool:
        return self.start <= other.end and other.start <= self.end


def no_beacon_count(sensors, beacons, target_y: int) -> int:
    ranges = []
    target_range = Range(target_y, target_y)
    for (x, y), d in sensors.items():
        if not Range(y - d, y + d).overlaps(target_range):
            continue
        y_dist = d - abs(y - target_y)
        ranges.append(Range(x - y_dist, x + y_dist))
    range = Range.union(*ranges)[0]
    object_xs = {
        x
        for x, y in [*sensors, *beacons]
        if y == target_y and range.overlaps(Range(x, x))
    }
    return len(range) - len(object_xs)


# https://www.reddit.com/r/adventofcode/comments/zmcn64/comment/j0b90nr/
def tuning_frequency(sensors, beacons, space: int) -> int:
    acoeffs, bcoeffs = set(), set()
    for (x, y), d in sensors.items():
        acoeffs.add(y - x + d + 1)
        acoeffs.add(y - x - d - 1)
        bcoeffs.add(x + y + d + 1)
        bcoeffs.add(x + y - d - 1)
    for a, b in product(acoeffs, bcoeffs):
        x, y = (b - a) // 2, (b + a) // 2
        if 0 < x < space and 0 < y < space:
            if all(mdist((x, y), t) > sensors[t] for t in sensors):
                return FREQUENCY_MULTIPLIER * x + y
    return -1


if __name__ == "__main__":
    main(
        lambda s, target_y: no_beacon_count(*parse(s), target_y),
        lambda s, space: tuning_frequency(*parse(s), space),
    )

File: ts/2022/day15.ts
import { main } from 'lib/advent';
import { PointMap, PointSet, distance } from 'lib/coords';
import { iter } from 'lib/iter';
import { lines } from 'lib/util';

type Input = { sensors: PointMap<number>; beacons: PointSet };
type Range = [number, number];
const FREQUENCY_MULTIPLIER = 4000000;

function parse(lines: string[]): Input {
  const pairs = lines.map((line) => {
    const [sx, sy, bx, by] = line.match(/-?\d+/g).slice(0, 4).map(Number);
    return { sensor: { x: sx, y: sy }, beacon: { x: bx, y: by } };
  });
  const sensors = new PointMap(
    pairs.map(({ sensor, beacon }) => [sensor, distance(sensor, beacon)])
  );
  const beacons = new PointSet(pairs.map(({ beacon }) => beacon));
  return { sensors, beacons };
}

function rangeOverlaps(a: Range, b: Range): boolean {
  return a[0] <= b[1] && b[0] <= a[1];
}

function unionRanges(ranges: [number, number][]): [number, number][] {
  const sorted = ranges.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
  const union = [];
  let last: [number, number];
  for (const [start, end] of sorted) {
    if (last && last[1] >= start - 1) {
      last[1] = Math.max(last[1], end);
    } else {
      last = [start, end];
      union.push(last);
    }
  }
  return union;
}

function noBeaconCount({ sensors, beacons }: Input, targetY: number) {
  const ranges = iter(sensors.entries())
    .filter(([sensor, d]) =>
      rangeOverlaps([sensor.y - d, sensor.y + d], [targetY, targetY])
    )
    .map(([sensor, d]) => {
      const yDistance = d - Math.abs(sensor.y - targetY);
      return [sensor.x - yDistance, sensor.x + yDistance] as Range;
    })
    .toArray();

  const union = unionRanges(ranges);
  const noBeacon = iter(union).map(([start, end]) => end - start + 1);
  const objects = iter([...sensors.keys(), ...beacons])
    .filter(({ y }) => y === targetY)
    .map(({ x }) => x)
    .uniq()
    .filter((x) => union.some((range) => rangeOverlaps(range, [x, x])));
  return noBeacon.sum() - objects.size;
}

// TODO: optimize
function tuningFrequency({ sensors, beacons }: Input, space: number) {
  const ranges = Array(space + 1)
    .fill(null)
    .map(() => []);

  for (const p of [...sensors.keys(), ...beacons]) {
    if (p.y >= 0 && p.y <= space) ranges[p.y].push([p.x, p.x]);
  }

  for (const [sensor, d] of sensors.entries()) {
    const top = sensor.y - d;
    const bottom = sensor.y + d;

    for (let y = Math.max(0, top); y <= Math.min(space, bottom); ++y) {
      const yDistance = d - Math.abs(sensor.y - y);
      ranges[y].push([sensor.x - yDistance, sensor.x + yDistance] as Range);
    }
  }

  for (let y = 0; y <= space; ++y) {
    const union = unionRanges(ranges[y]);
    if (union.length > 1) {
      const x = union[0][1] + 1;
      return x * FREQUENCY_MULTIPLIER + y;
    }
  }
}

main(
  (s, { target_y }) => noBeaconCount(parse(lines(s)), target_y as number),
  (s, { space }) => tuningFrequency(parse(lines(s)), space as number)
);
